### ISSUE TITLE
[Performance] Potential regression in 1.0.0-Beta1

### ISSUE THREAD

**Issue by nknize** [opensearch-project/OpenSearch#673]
**Describe the issue**
Benchmarking has been set up and the following numbers captured today:

| Product/ Version       | <br>Architecture | Description      | Instance Type | Workload Details                                 | Index Latency (ms) - p50 | **Index Latency (ms) -** p90 | **Index Latency (ms) -** p99 | **Index Latency (ms) -** p100 | **Index Throughput (docs/s) -** p0 | **Index Throughput (docs/s) -** p50 | **Index Throughput (docs/s) -** p100 | Ops Count | Op Error Count | Error Rate | Query Latency (ms) - p50 | Query Latency (ms) - p90 | Query Latency (ms) - p99 | Query Latency (ms) - p100 | Query Throughput (docs/s) - p0 | Query Throughput (docs/s) - p50 | Query Throughput (docs/s) - p100 | Ops Count | Op Error Count | Error Rate | CPU (%) - p50 | CPU (%) - p90 | CPU (%) - p99 | CPU (%) - p100 | Memory (%) - p50 | Memory (%) - p90 | Memory (%) - p99 | Memory (%) - p100 | GC (ms) - Old | GC (ms) - Young |
| ---------------------- | ---------------- | ---------------- | ------------- | ------------------------------------------------ | ------------------------ | ---------------------------- | ---------------------------- | ----------------------------- | ---------------------------------- | ----------------------------------- | ------------------------------------ | --------- | -------------- | ---------- | ------------------------ | ------------------------ | ------------------------ | ------------------------- | ------------------------------ | ------------------------------- | -------------------------------- | --------- | -------------- | ---------- | ------------- | ------------- | ------------- | -------------- | ---------------- | ---------------- | ---------------- | ----------------- | ------------- | --------------- |
| ODFE 1.13.2            | X64              | With security    | m5.xlarge     | nyc\_taxis/ 2 warmupIterations/ 3 testIterations | 2,371.4                  | 3,077.4                      | 4,294.3                      | 9,076.4                       | 32,124.7                           | 33,120.6                            | 38,174.6                             | 46,871    | \-             | \-         | 175.2                    | 203.5                    | 216.6                    | 273.3                     | 1.748                          | 1.753                           | 1.767                            | 1,510     | \-             | \-         | 97            | 97            | 98            | 98             | 50               | 64               | 67               | 70                | \-            | 1,678,812       |
| OpenSearch-1.0.0-beta1 | X64              | With security    | m5.xlarge     | nyc\_taxis/ 2 warmupIterations/ 3 testIterations | 3,198.3                  | 4,432.5                      | 7,910.1                      | 10,011.5                      | 22,784.5                           | 23,332.2                            | 27,295.6                             | 47,649    | 58             | 0.10%      | 261                      | 305.1                    | 381.9                    | 411.6                     | 2.009                          | 2.015                           | 2.031                            | 906       | \-             | \-         | 98            | 98            | 98            | 98             | 51               | 64               | 68               | 70                | \-            | 1,794,217       |
| ODFE 1.13.2            | X64              | Without security | m5.xlarge     | nyc\_taxis/ 2 warmupIterations/ 3 testIterations | 1,915.7                  | 2,592.7                      | 3,620.7                      | 6,655.2                       | 39,218.4                           | 40,990.2                            | 47,736.1                             | 46,193    | \-             | \-         | 266.5                    | 280.5                    | 289.5                    | 294                       | 1.747                          | 1.751                           | 1.764                            | 1,510     | \-             | \-         | 98            | 98            | 99            | 100            | 48               | 62               | 67               | 75                | \-            | 1,281,143       |
| OpenSearch-1.0.0-beta1 | X64              | Without security | m5.xlarge     | nyc\_taxis/ 2 warmupIterations/ 3 testIterations | 2,986                    | 4,170.5                      | 7,673.3                      | 13,117.2                      | 24,271.4                           | 25,024.7                            | 29,637.7                             | 47,475    | \-             | \-         | 316.7                    | 335.4                    | 424.2                    | 466.2                     | 2.002                          | 2.017                           | 2.033                            | 906       | \-             | \-         | 99            | 99            | 99            | 99             | 49               | 62               | 67               | 69                | \-            | 1,543,353       |

Initial results indicate a possible regression. Further root cause analysis, from benchmark server configuration to investigating dependency performance, needs to be done to first confirm these as valid performance regression indicators. If the behavior is confirmed the regression will be tracked down and a patch will be opened and merged in advance of the Beta 2 release. This issue is to track the progress and transparently communicate the update as it becomes available.

**Comment by CEHENKLE** [comment]
Tl;Dr  The original large performance degradation we saw was caused by an issue in the performance tool.  Now that we've cleared that out, we're still seeing a (smaller) performance degradation, but we've got a smoking gun that the version changes are causing the issue.   So we believe this is ultimately related to #693, and should be resolved with that issue. 

We've been digging into this, and here's what we've found:

The results above are from an internal tool based on Rally.  We had previously run Rally against OpenSearch Alpha, so in order to get an Apples to Apples comparison, we wanted to rerun Rally against OpenSearch Beta.   We did that and saw similar results:
[osbeta1_0_0_vs_es_&_10_2_nyc_taxi.pdf](https://github.com/opensearch-project/OpenSearch/files/6480053/osbeta1_0_0_vs_es_._10_2_nyc_taxi.pdf)

BUT!  (and the plot thickens!)
Then we figure out what was going on with Rally and the internal tool.  Rally picks the branch of the track based on the version reported by the cluster, for ODFE that’s ‘7.10.2’, for OpenSearch that’s ‘1.0.0-beta1’.  This means that Rally used the 5.0 branch of nyc-taxis when running OpenSearch instead of the 7.1 branch that it used for ODFE.  So from out POV, that was (relatively) good news since it meant that these results have no meaning, since the track and operations were different.  

BUT!  (more plot!)
Once we'd figure out a way to get Rally to use the same tracks we ran again... and we still seeing some degradation.  But not nearly as bad.  

<html>
<body>


Product/ Version | Architecture | Description | Instance Type | Workload Details | Index | Query | CPU (%) |   |   |   | Memory (%) |   |   |   | GC (ms) |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Latency (ms) | Throughput (docs/s) | Operation Counts | Latency (ms) | Throughput (docs/s) |   |   | Operation Counts |   |   |   |   |   |   |   |   |   |   |   |  
p50 | p90 | p99 | p100 | p0 | p50 | p100 | Ops Count | Op Error Count | Error Rate | p50 | p90 | p99 | p100 | p0 | p50 | p100 | Ops Count | Op Error Count | Error Rate | p50 | p90 | p99 | p100 | p50 | p90 | p99 | p100 | Old | Young
ODFE 1.13.2 | X64 | With security | m5.xlarge | nyc_taxis/ 2 warmupIterations/ 3 testIterations | 2,371.4 | 3,077.4 | 4,294.3 | 9,076.4 | 32,124.7 | 33,120.6 | 38,174.6 | 46,871 | - | - | 175.2 | 203.5 | 216.6 | 273.3 | 1.748 | 1.753 | 1.767 | 1,510 | - | - | 97 | 97 | 98 | 98 | 50 | 64 | 67 | 70 | - | 1,678,812
OpenSearch-1.0.0-beta1 | X64 | With security | m5.xlarge | nyc_taxis/ 2 warmupIterations/ 3 testIterations | 2,530.2 | 3,251.7 | 4,833.3 | 7,956.3 | 30,336.8 | 31,283.1 | 36,216.9 | 47,046 | - | - | 178.3 | 206.9 | 221.9 | 225.2 | 1.749 | 1.754 | 1.767 | 1,510 | - | - | 97 | 97 | 98 | 98 | 53.5 | 64 | 68 | 70 | - | 1,781,648
Comparison |   |   |   |   | -6.70% | -5.66% | -12.55% | 12.34% | -5.57% | 5.55% | 5.13% |   |   |   | -1.77% | -1.67% | -2.45% | 17.60% | 0.06% | 0.06% | 0.00% |   |   |   |   |   |   |   |   |   |   |   |   |  
ODFE 1.13.2 | X64 | Without security | m5.xlarge | nyc_taxis/ 2 warmupIterations/ 3 testIterations | 1,915.7 | 2,592.7 | 3,620.7 | 6,655.2 | 39,218.4 | 40,990.2 | 47,736.1 | 46,193 | - | - | 266.5 | 280.5 | 289.5 | 294 | 1.747 | 1.751 | 1.764 | 1,510 | - | - | 98 | 98 | 99 | 100 | 48 | 62 | 67 | 75 | - | 1,281,143
OpenSearch-1.0.0-beta1 | X64 | Without security | m5.xlarge | nyc_taxis/ 2 warmupIterations/ 3 testIterations | 2,252 | 2,984 | 4,472.3 | 7,873 | 33,929.4 | 35,021.5 | 41,034.1 | 46,697 | - | - | 204.4 | 217.9 | 226.7 | 235 | 1.749 | 1.754 | 1.768 | 1,510 | - | - | 98 | 98 | 99 | 99 | 48 | 62 | 66 | 68 | - | 1,328,941
Comparison |   |   |   |   | -17.55% | -15.09% | -23.52% | -18.30% | -13.49% | -14.56% | -14.04% |   |   |   | 23.30% | 22.32% | 21.69% | 20.07% | 0.11% | 0.17% | 0.23% |   |   |   |   |   |   |   |   |   |   |   |   |  


</body>
</html>

BUT! (still more!)
So if we believe (and we do) that the original performance impact was because we was using different tracks, where was the new issue coming from?  @harold-wang has been digging in, running tests in rally with different builds, removing different PRs.  We have definitely gotten a smoking gun that the issue started with our version changes.   We have a couple of theories on what's going on that's causing the degradation, but I'll let @harold-wang post more about it.

Thanks very much to @harold-wang, @cmanning09, @nknize  and Phil Treddenick for digging in.

**Comment by nknize** [comment]
Initial [PR has been opened](https://github.com/opensearch-project/OpenSearch/pull/708) to fix BWC for legacy transport clients.

**Comment by harold-wang** [comment]
Performance result is gone after I change the version id from 1.0.0 to 7.10.2 in ESRally test.

| Min Throughput | Mean Throughput | Median Throughput | Max Throughput | 50th percentile latency | 90th percentile latency | 99th percentile latency | 99.9th percentile latency | 99.99th percentile latency | 100th percentile latency | percentile service time | percentile service time | percentile service time | percentile service time | percentile service time | percentile service time
|------------ | -------------|------------ | -------------|------------ | -------------|------------ | -------------|------------ | -------------|------------ | -------------|------------ | -------------|------------ | -------------|

**Comment by harold-wang** [comment]
[OpenSearch100_main_vs_elasticsearch_7_10_3_performanceIssue_fakeversionfix.pdf](https://github.com/opensearch-project/OpenSearch/files/6609448/OpenSearch100_main_vs_elasticsearch_7_10_3_performanceIssue_fakeversionfix.pdf)

**Comment by harold-wang** [comment]
Re-open this issue, we are running the same test by Mensor against the cluster with opensearch BWC enabled. will update. 

https://prod.mensor.searchservices.aws.dev/test/26010c94-647f-466b-a386-e23c370f1345
https://search-mensor-metrics-jz77bwnyqqiekf7g2fmkfmfncu.eu-west-1.es.amazonaws.com/_plugin/kibana/app/kibana?security_tenant=global#/dashboard/2366e970-419d-11eb-ac43-fb82b7859827?_g=(filters:!(),refreshInterval:(display:'1%20minute',pause:!f,value:60000),time:(from:now-24h,to:now))&_a=(description:'',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!f),panels:!((embeddableConfig:(title:''),gridData:(h:15,i:'5ef0a36f-ec97-4d52-96d1-6ccee5c2adce',w:48,x:0,y:0),id:ccb1ec30-419f-11eb-9fed-f3dfeb55b7ea,panelIndex:'5ef0a36f-ec97-4d52-96d1-6ccee5c2adce',type:visualization,version:'7.7.0'),(embeddableConfig:(title:''),gridData:(h:8,i:'70420d98-9111-4f74-ab57-48dad5a4200a',w:32,x:0,y:15),id:'27b72960-41a0-11eb-94ff-69ed4c039cfc',panelIndex:'70420d98-9111-4f74-ab57-48dad5a4200a',type:visualization,version:'7.7.0'),(embeddableConfig:(title:''),gridData:(h:8,i:'1b441e91-d728-4344-b5d5-e48d7b482328',w:16,x:32,y:15),id:'9df6dd90-43c2-11eb-94ff-69ed4c039cfc',panelIndex:'1b441e91-d728-4344-b5d5-e48d7b482328',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:15,i:b82abd45-c77e-42c2-a138-e049aed07172,w:48,x:0,y:23),id:bf5343d0-5528-11eb-b91d-d9b60acd3a47,panelIndex:b82abd45-c77e-42c2-a138-e049aed07172,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:fa7c7977-37a2-4bf7-b9f4-73711e5243cd,w:48,x:0,y:38),id:d07659f0-416d-11eb-9fed-f3dfeb55b7ea,panelIndex:fa7c7977-37a2-4bf7-b9f4-73711e5243cd,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'9df12f96-8cb2-4329-837b-33dbc31e2d14',w:48,x:0,y:49),id:f2231c00-416d-11eb-9fed-f3dfeb55b7ea,panelIndex:'9df12f96-8cb2-4329-837b-33dbc31e2d14',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'1d998eca-9033-48da-b383-944cd9c9f0f7',w:48,x:0,y:60),id:'2061ab60-5a99-11eb-94ff-69ed4c039cfc',panelIndex:'1d998eca-9033-48da-b383-944cd9c9f0f7',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:b59878b5-418e-4e62-854a-182d3f527ef9,w:24,x:0,y:71),id:'5eeadb70-416e-11eb-9fed-f3dfeb55b7ea',panelIndex:b59878b5-418e-4e62-854a-182d3f527ef9,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'337414c5-61ce-496f-8f4d-b43f40aaffbb',w:24,x:24,y:71),id:a4b02f70-416e-11eb-abea-1752399bd867,panelIndex:'337414c5-61ce-496f-8f4d-b43f40aaffbb',type:visualization,version:'7.7.0'),(embeddableConfig:(title:''),gridData:(h:7,i:'8d58abfe-ff93-431c-923b-0db04ea9ec3a',w:48,x:0,y:82),id:c8324220-5aa6-11eb-abea-1752399bd867,panelIndex:'8d58abfe-ff93-431c-923b-0db04ea9ec3a',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:fd9ce5e4-495c-4ed3-b0e4-f5e40c234b14,w:24,x:0,y:89),id:'9b6e1000-4176-11eb-abea-1752399bd867',panelIndex:fd9ce5e4-495c-4ed3-b0e4-f5e40c234b14,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'4870dc1c-5da5-4a3e-8d1c-e43a23c54811',w:24,x:24,y:89),id:e7cb8a60-4174-11eb-b91d-d9b60acd3a47,panelIndex:'4870dc1c-5da5-4a3e-8d1c-e43a23c54811',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:cb980b53-f209-487e-839e-d39ffca937a6,w:24,x:0,y:100),id:'6acd06e0-4176-11eb-b91d-d9b60acd3a47',panelIndex:cb980b53-f209-487e-839e-d39ffca937a6,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'76f0ec10-627b-4ace-9eb7-9f02bf05d8cc',w:24,x:24,y:100),id:'7f9ac2c0-4175-11eb-94ff-69ed4c039cfc',panelIndex:'76f0ec10-627b-4ace-9eb7-9f02bf05d8cc',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'4d7508d5-245d-482f-ab6b-df44a3aa2f11',w:24,x:0,y:111),id:abd0a990-4175-11eb-9fed-f3dfeb55b7ea,panelIndex:'4d7508d5-245d-482f-ab6b-df44a3aa2f11',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:b4133175-32ca-4703-9824-382c6d2b8758,w:24,x:24,y:111),id:'1530acf0-4176-11eb-94ff-69ed4c039cfc',panelIndex:b4133175-32ca-4703-9824-382c6d2b8758,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:ee172f65-1d1a-4e43-b391-3b38b0d74955,w:24,x:0,y:122),id:'3f86d770-419b-11eb-b91d-d9b60acd3a47',panelIndex:ee172f65-1d1a-4e43-b391-3b38b0d74955,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:f534741a-b7cc-4454-ac81-41e3a2aa4f76,w:24,x:24,y:122),id:'476cef80-4176-11eb-9fed-f3dfeb55b7ea',panelIndex:f534741a-b7cc-4454-ac81-41e3a2aa4f76,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'6fd55884-b028-4e27-b637-a7daab98848e',w:24,x:0,y:133),id:a76bac80-419b-11eb-b91d-d9b60acd3a47,panelIndex:'6fd55884-b028-4e27-b637-a7daab98848e',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'1efa0a4f-96fe-4fab-8d88-86897bb52851',w:24,x:24,y:133),id:'64a91130-419b-11eb-ac43-fb82b7859827',panelIndex:'1efa0a4f-96fe-4fab-8d88-86897bb52851',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'19b7c63b-ff40-4de4-99ce-7a1f42aef388',w:24,x:0,y:144),id:d7e85a30-419a-11eb-b91d-d9b60acd3a47,panelIndex:'19b7c63b-ff40-4de4-99ce-7a1f42aef388',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:cfc5116f-a02b-4e29-8e96-b31c95cdc52b,w:24,x:24,y:144),id:'16eed7e0-419b-11eb-94ff-69ed4c039cfc',panelIndex:cfc5116f-a02b-4e29-8e96-b31c95cdc52b,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'1d9b208e-635d-4ef3-8dff-589d0a8094e8',w:24,x:0,y:155),id:'67fdf570-4183-11eb-ac43-fb82b7859827',panelIndex:'1d9b208e-635d-4ef3-8dff-589d0a8094e8',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:f9feeb80-8600-4788-a9d3-5fbc1bd77a40,w:24,x:24,y:155),id:'83bf70f0-419b-11eb-abea-1752399bd867',panelIndex:f9feeb80-8600-4788-a9d3-5fbc1bd77a40,type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'4b0e744c-d5da-4fcc-9a69-31a23e196757',w:24,x:0,y:166),id:'85cb1f40-4176-11eb-abea-1752399bd867',panelIndex:'4b0e744c-d5da-4fcc-9a69-31a23e196757',type:visualization,version:'7.7.0'),(embeddableConfig:(),gridData:(h:11,i:'9aa13877-a0e3-4f39-a582-b356b75014ce',w:24,x:24,y:166),id:c7a68f10-419b-11eb-94ff-69ed4c039cfc,panelIndex:'9aa13877-a0e3-4f39-a582-b356b75014ce',type:visualization,version:'7.7.0')),query:(language:kuery,query:%2226010c94-647f-466b-a386-e23c370f1345%22),timeRestore:!f,title:'Mensor%20Test%20Run',viewMode:view)

**Comment by harold-wang** [comment]
Summary:

1. OpenSearch shows higher CPU usage in P50, P90 and P99 (23% to 25%)
2. OpenSearch shows lower memory usage(-15% to -25%)
3. Query throughput is almost same
4. OpenSearch has 16% higher index throughput
5. OpenSearch has 50% higher Query latency in P50 and P99, P99, P100

**Comment by dblock** [comment]
@harold-wang what was the resolution here?

**Comment by nknize** [comment]
I think this was resolved as a versioning discrepancy due to the versioning framework using old versions of elasticsearch? bumping to v.1.2.0 but I think this can be closed?
